src/rust/lib/graph-editor/src/component/node/port.rs

```diff
@@ -292,12 +324,52 @@ pub struct Port<T:PortShapeViewDefinition> {
     pub data : Rc<PortData<T>>
 }
 
+/// Weak version of `Port`.
+#[derive(Clone,CloneRef,Debug)]
+pub struct WeakPort<T:PortShapeViewDefinition+Clone> {
+    data : Weak<PortData<T>>
+}
+
+impl<T:PortShapeViewDefinition> StrongRef for Port<T> {
+    type WeakRef = WeakPort<T>;
+    fn downgrade(&self) -> WeakPort<T> {
+        WeakPort {data:Rc::downgrade(&self.data)}
+    }
+}
+
+impl<T:PortShapeViewDefinition> WeakRef for WeakPort<T> {
+    type StrongRef = Port<T>;
+    fn upgrade(&self) -> Option<Port<T>> {
+        self.data.upgrade().map(|data| Port{data})
+    }
+}
+
 /// Type that represents an input port.
 pub type InputPort = Port<InputPortView>;
 
 /// Type that represents an output port.
 pub type OutputPort = Port<OutputPortView>;
 
+/// Enum over all possible port types.
+#[derive(Clone,CloneRef,Debug)]
+#[allow(missing_docs)]
+pub enum IOPort{
```

Comment:
This is not proper name in rust terms. I hate it too, but it should be `IoPort`. Maybe is there a better name?

Reply:
Maybe this should even be the main `Port` struct. Need to consider the naming and implications.